### PR TITLE
feat(backend/frontend/shared): Add notes to tasks bb-419

### PR DESCRIPTION
**Pull request by JD-Rosales**
### Summary

- Added POST /task/notes to create a notes for a task.
- Added GET /task/note/:id to get all notes for a task.
- Added  new migration for task note table.
- Added  model, entity, and repository for task note.
- Added validation

### CODE REVIEW

**Comment by JD-Rosales** [comment]
This is just the backend part. I'll work on the frontend once the API is approved.

File: apps/backend/src/db/migrations/20240918040515_add_task_notes_table.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import { type Knex } from "knex";
+
+const TableName = {
+	NOTES: "task_notes",
```
Comment (fshabanov) [inline apps/backend/src/db/migrations/20240918040515_add_task_notes_table.ts:4]:
```suggestion
	TASK_NOTES: "task_notes",
```

File: apps/backend/src/db/migrations/20240918040515_add_task_notes_table.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+} as const;
+
+const DELETE_STRATEGY = "CASCADE";
+
+async function up(knex: Knex): Promise<void> {
+	await knex.schema.createTable(TableName.NOTES, (table) => {
+		table.increments(ColumnName.ID).primary();
+		table
+			.integer(ColumnName.TASK_ID)
+			.unsigned()
```
Comment (fshabanov) [inline apps/backend/src/db/migrations/20240918040515_add_task_notes_table.ts:23]:
```suggestion
```
We don't need it

File: apps/backend/src/modules/tasks/task.controller.ts [hunk 3]
```diff
@@ -102,6 +106,74 @@ class TaskController extends BaseController {
...
+	 *             schema:
+	 *               $ref: "#/components/schemas/CommonErrorResponse"
+	 *       404:
+	 *         description: Task not found
+	 *         content:
+	 *           application/json:
+	 *             schema:
+	 *               $ref: "#/components/schemas/CommonErrorResponse"
+	 */
+	private async addNote(
```
Comment (fshabanov) [inline apps/backend/src/modules/tasks/task.controller.ts:168]: You should also add a possibility to get notes

File: apps/backend/src/modules/tasks/task.service.ts [hunk 4]
```diff
@@ -51,6 +53,21 @@ class TaskService implements Service {
 		this.taskRepository = taskRepository;
 	}
 
+	public async addNote(payload: TaskNoteRequestDto): Promise<TaskNoteDto> {
+		const task = await this.taskRepository.find(payload.taskId);
+
+		if (!task) {
+			throw new TaskError({
+				message: ErrorMessage.TASK_NOT_FOUNT,
```
Comment (fshabanov) [inline apps/backend/src/modules/tasks/task.service.ts:61]:
```suggestion
				message: ErrorMessage.TASK_NOT_FOUND,
```

File: packages/shared/src/modules/tasks/libs/enums/task-note-validation-message.enum.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+const TaskNoteValidationMessage = {
+	INVALID_TASK_ID: "Task ID must be a valid number.",
+	REQUIRED_CONTENT: "Note content is required.",
+	REQUIRED_TASK_ID: "Task ID is required.",
```
Comment (fshabanov) [inline packages/shared/src/modules/tasks/libs/enums/task-note-validation-message.enum.ts:4]: Messages should be user-friendly

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Comment by fshabanov** [comment]
Please, start working on Frontend, you don't need to wait for a Backend to be approved to start

File: apps/backend/src/modules/tasks/task.controller.ts [hunk 6]
```diff
@@ -102,6 +107,85 @@ class TaskController extends BaseController {
...
+		});
+
+		this.addRoute({
+			handler: (options) =>
+				this.getNotesByTaskId(
+					options as APIHandlerOptions<{
+						params: TaskNoteParametersDto;
+					}>,
+				),
+			method: "GET",
```
Comment (fshabanov) [inline apps/backend/src/modules/tasks/task.controller.ts:130]: Add a preHandler to check if the user gets notes for their own tasks

File: apps/backend/src/modules/tasks/task.service.ts [hunk 7]
```diff
@@ -104,6 +121,23 @@ class TaskService implements Service {
...
 
+	public async getNotesByTaskId(taskId: number): Promise<TaskNoteDto[]> {
+		const task = await this.taskRepository.find(taskId);
+
+		if (!task) {
+			throw new TaskError({
+				message: ErrorMessage.TASK_NOT_FOUND,
+				status: HTTPCode.NOT_FOUND,
+			});
+		}
```
Comment (fshabanov) [inline apps/backend/src/modules/tasks/task.service.ts:132]: I think this check is redundant, if there's no task, then the notes array will be empty

File: packages/shared/src/modules/tasks/libs/enums/task-note-validation-message.enum.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+const TaskNoteValidationMessage = {
+	INVALID_TASK_ID: "Task ID must be a valid number.",
```
Comment (fshabanov) [inline packages/shared/src/modules/tasks/libs/enums/task-note-validation-message.enum.ts:2]: This message is not very user-friendly

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by marharita08 — COMMENTED** [review]
1. Does this task includes attaching images to note?
2. As for me it would be better to create separate module (including repository, service and controller) for notes.

**Comment by JD-Rosales** [comment]
> 1. Does this task includes attaching images to note?
> 2. As for me it would be better to create separate module (including repository, service and controller) for notes.

1. No, for now only a note.
2. As for me I prefer it within task module because of their common relationship and related function to avoid unnecessary dependency between them.

File: apps/backend/src/modules/tasks/task.controller.ts (hunk 6 above)
Comment (JD-Rosales) [reply apps/backend/src/modules/tasks/task.controller.ts:130]: There's already a preHandler hook to check if the user has access to task when updating a task. Is it okay if I reuse that?

`const checkAccessToTask =
	(taskService: TaskService): APIPreHandler =>
	async (request) => {
		const { params, user } = request;

		const taskId = (params as TaskUpdateParametersDto).id;

		const task = await taskService.find(taskId);

		if ((user as UserDto).id !== task?.userId) {
			throw new TaskError({
				message: ErrorMessage.FORBIDDEN,
				status: HTTPCode.FORBIDDEN,
			});
		}
	};
`

**Comment by JD-Rosales** [comment]
Should I also create a Textarea component for this task? It could also be used when updating task descriptions and note content.

**Review by fshabanov — APPROVED** [review]
Backend looks good, waiting for Frontend

**Pull request merged**